### PR TITLE
Rework error handling within the ASP logic program

### PR DESCRIPTION
**Pull request by alalazo**
This PR reworks the rules that determine errors in the logic program to be based on cardinality constraints and to have a single optimization criterion instead of three. This speeds-up the concretizer generally, while it keeps the previous model for solvable specs.

Modifications:
- [x] Reduce effort on grounding by employing cardinality constraints (6d16ce291701f52204c6966fb45225032794727b)
- [x] Display all errors from concretization, instead of just one (e477119af9f65e493abdc6c34c2ff4aa0cf6bd83)
- [x] Use a single optimization criteria for errors (92f9500789084a91a4ae2beaead81115a4273b63)
- [x] Differentiate errors according to weights (7dec96ea851692e6c791c7592c0f8c952449b1c1)

### CODE REVIEW

**Comment by alalazo** [comment]
Data for benchmarking clingo:

[radiuss.txt](https://github.com/spack/spack/files/11105282/radiuss.txt) (abstract specs solved)
[radiuss_develop.csv](https://github.com/spack/spack/files/11105281/radiuss_develop.csv) (develop at 781c87840d1c95c3cfce5e046d9b683a29d67a39)
[radiuss_error_multiple.csv](https://github.com/spack/spack/files/11105280/radiuss_error_multiple.csv) (this PR at e477119af9f65e493abdc6c34c2ff4aa0cf6bd83)
[radiuss_error_single.csv](https://github.com/spack/spack/files/11105279/radiuss_error_single.csv) (this PR at 92f9500789084a91a4ae2beaead81115a4273b63)

Comparing this PR against `develop`:

totals

The effect of squeezing all the error criteria into a single criterion is the following:

totals

**Comment by alalazo** [comment]
Here's the total concretization time on `develop` vs. the PR with a single criterion:

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.3754 |                  14.681  |      10.3472  |
| ascent        |                       33.4807 |                  30.9698 |       7.49971 |
| axom          |                       33.0372 |                  30.7192 |       7.01629 |
| blt           |                       15.3174 |                  13.9308 |       9.05215 |
| caliper       |                       16.7042 |                  15.1411 |       9.35787 |
| care          |                       19.0532 |                  15.7361 |      17.4094  |
| chai          |                       17.8564 |                  14.9314 |      16.3807  |
| conduit       |                       17.1218 |                  15.5197 |       9.35714 |
| dihydrogen    |                       17.3922 |                  15.6578 |       9.97219 |
| flux-core     |                       16.3239 |                  14.7064 |       9.90911 |
| flux-sched    |                       16.6058 |                  14.8185 |      10.7631  |
| glvis         |                       32.8462 |                  30.2662 |       7.85493 |
| hydrogen      |                       17.0082 |                  15.3271 |       9.8841  |
| hypre         |                       17.964  |                  16.3648 |       8.90227 |
| lbann         |                       34.1094 |                  31.1512 |       8.67281 |
| lvarray       |                       23.5327 |                  20.0112 |      14.9643  |
| mfem          |                       30.3318 |                  27.9654 |       7.80181 |
| py-hatchet    |                       20.2211 |                  18.1413 |      10.2855  |
| py-maestrowf  |                       16.8339 |                  15.5212 |       7.79791 |
| py-merlin     |                       19.8586 |                  17.1467 |      13.6561  |
| py-shroud     |                       15.1449 |                  14.2487 |       5.91792 |
| raja          |                       17.479  |                  14.4169 |      17.5189  |
| samrai        |                       15.8138 |                  14.0608 |      11.0854  |
| scr           |                       16.8897 |                  14.9935 |      11.2269  |
| sundials      |                       27.337  |                  24.254  |      11.2779  |
| umpire        |                       17.1102 |                  13.9857 |      18.2609  |
| visit         |                       36.1799 |                  31.4682 |      13.0231  |
| xbraid        |                       15.7402 |                  14.2889 |       9.2201  |
| zfp           |                       15.3423 |                  14.1555 |       7.73545 |

**Comment by alalazo** [comment]
Finally, the error messages now look like (here I added a `depends_on` requirement on purpose to libpng):

Screenshot from 2023-03-29 17-31-56

File: lib/spack/spack/solver/concretize.lp [hunk 1]
```diff
@@ -77,21 +95,11 @@ version_satisfies(Package, Constraint, HashVersion) :- version_satisfies(Package
...
-error(2, "No versions available for package '{0}'", Package)
-  :- attr("node", Package), not attr("version", Package, _).
 
 % A virtual package may or may not have a version, but never has more than one
-error(2, "No version for '{0}' satisfies '@{1}' and '@{2}'", Virtual, Version1, Version2)
+error(2, "Cannot select a single version for virtual '{0}'", Virtual)
   :- attr("virtual_node", Virtual),
-     attr("version", Virtual, Version1),
-     attr("version", Virtual, Version2),
-     Version1 < Version2. % see[1]
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp]: @becker33 @tgamblin To answer the question from yesterday, on how can a cardinality rule be faster than comparing each pair with each other, I think this slide (from [here](https://github.com/potassco-asp-course/course/releases/download/v1.4.1/language.pdf)) helps:
Screenshot from 2023-03-30 16-11-19

The cardinality constraint is implemented in the form of a counter (`x(i, j)` in the slide above) which allows a lot of "fast tracks" and simplifications, if compared to using all the combinations of possible versions.

**Comment by alalazo** [comment]
@becker33 I added back weights, and confirm the benchmark is not affected:

[radiuss_weights.csv](https://github.com/spack/spack/files/11138142/radiuss_weights.csv) (7dec96ea851692e6c791c7592c0f8c952449b1c1)
[radiuss_no_weights.csv](https://github.com/spack/spack/files/11138144/radiuss_base.csv) (020bfc0feb429f6a4de35d6604ab98e621ef2262)
[radiuss_develop.csv](https://github.com/spack/spack/files/11138145/radiuss_develop.csv)

File: lib/spack/spack/solver/asp.py [hunk 2]
```diff
@@ -586,6 +586,54 @@ def extract_args(model, predicate_name):
...
+        msg = msg.format(*args)
+        specs_to_construct = re.findall(r"Spec\(([^)]*)\)", msg)
+        for spec_str in specs_to_construct:
+            msg = msg.replace("Spec(%s)" % spec_str, str(spack.spec.Spec(spec_str)))
+
+        return msg
+
+    def message(self, errors) -> str:
+        messages = [
+            f"  {idx+1: 2}. {self.handle_error(msg, *args)}"
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py:620]: Why are we enforcing a column width here? I don't see this lining up with the outputs you've put in the discussion so far?

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/solver/asp.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/solver/asp.py:620]: I mixed up string padding and integer padding, I'll mark this resolved.

**Comment by alalazo** [comment]
Thanks @becker33 !

**Pull request merged**